phac-nml/irida-next
# Add job and task for attachment cleanup

## What does this PR do and why?
_Describe in detail what your merge request does and why._
Fixes #220 
Created a new job `AttachmentsCleanupJob` which accepts 1 argument `days_old` and deletes (destroys) attachments that were soft deleted when they are at least `days_old` old.
Created a new task `rake "attachments:cleanup[X]" which calls the above job where X is `days_old`

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._
![image](https://github.com/phac-nml/irida-next/assets/16961365/d4f292bc-d6f1-4e68-97ad-4fad00e9af74)


## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._
Create a clean database
```bash
scripts/clean_dev.sh
```
Run IRIDA Next, Sign in and add 3 files to a sample
![image](https://github.com/phac-nml/irida-next/assets/16961365/f75ea4df-aa97-493b-8ae4-beecc6e8f76c)
Run a rails console
```bash
rails c
```
view a list of attachment id's with
```ruby
Attachment.all.map(&:id)
```
You should have 3 attachment with ids `[1, 2, 3]`
![image](https://github.com/phac-nml/irida-next/assets/16961365/ac3eff1e-c028-4d9d-a799-155590fd9c1f)
delete a file via the ui
![image](https://github.com/phac-nml/irida-next/assets/16961365/45bd399e-9d70-4a4b-82a4-d52894b59ce5)
you can verify deletion with
```ruby
# to see files not deleted
Attachment.all.map(&:id)
# to see deleted files
Attachment.only_deleted.map(&:id)
# to see how many actual files exist since deleted files are only soft deleted
ActiveStorage::Attachment.count
```
![image](https://github.com/phac-nml/irida-next/assets/16961365/a0e1b3f0-8485-4974-8f66-93f8463cbe4d)

~Invent time travel and go some time into the future~
Update the deleted attachments `deleted_at` field such that it was deleted 10 days ago. Then verify it was updated.
```ruby
Attachment.only_deleted[0].update(deleted_at: Attachment.only_deleted[0].deleted_at - 10.days)
```
![image](https://github.com/phac-nml/irida-next/assets/16961365/751fd955-0e1f-4928-8b5c-2fa669a61592)

run the task (bash, not ruby terminal)
```bash
rake attachments:cleanup
```
verify the file was hard deleted (rails console)
```ruby
Attachment.all.map(&:id)
Attachment.only_deleted.map(&:id)
ActiveStorage::Attachment.count
```
![image](https://github.com/phac-nml/irida-next/assets/16961365/dd536e59-c880-4e7d-ab40-561b6ec50b07)

Additional tests
* deletion of multiple attachments
* setting the number of days instead of taking the default of 7
  * example: `rake "attachments:cleanup[3]`
  * specifying the number of days always uses the previous day at midnight as the cutoff to give some additional time for safety.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
